src/main/scala/scalaz/reactive/Future.scala

```diff
@@ -2,33 +2,44 @@ package scalaz.reactive
 
 import scalaz.Scalaz._
 import scalaz.reactive.Time.{ NegInf, PosInf }
+import scalaz.zio.IO
 import scalaz.{ Applicative, Functor, Monad, Monoid }
 
-case class Future[+A](time: () => Time, force: () => A) {
+object Types {
+  type Infallible[A] = IO[Nothing, A]
+}
+
+case class Future[+A](force: IO[Nothing, (Time, A)]) {
 
   def + [AA >: A](other: Future[AA]): Future[AA] =
-    Future(
-      () => time().min(other.time()),
-      () => if (time() <= other.time()) force() else other.force()
-    )
+    Future(force.flatMap {
+      case (t1, a1) =>
+        other.force.map {
+          case (t2, a2) => if (t1 <= t2) (t1, a1) else (t2, a2)
+        } //There should be IO.ap, is there?
+    })
 
   def map[B](f: A => B): Future[B] =
-    Future(() => time(), () => f(force()))
+    Future(force.map { case (t, a) => (t, f(a)) })
 
   def ap[B](f: => Future[A => B]): Future[B] =
-    Future(() => time().max(f.time()), () => f.force()(force()))
+    Future(force.flatMap {
+      case (t, a) =>
+        f.force
+          .map { case (t2, f) => (t2.max(t), f(a)) } // There should be IO.ap, is there?
```

Comment:
Yes we could use IO ap instead of flatMap, I am not sure what are the difference in the semantics of the IO ap and flatMap, they are the same if ap is defined in terms of flatMap